### PR TITLE
Fix a bug in balancing training/valid/test in graph partitioning.

### PR DESCRIPTION
**Pull request by zheng-da**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/utils.py [hunk 1]
```diff
@@ -941,17 +941,32 @@ def partition_graph(g, node_data, edge_data, graph_name, num_partitions, output_
...
+            num_valid = th.sum(node_data[ntype]["val_mask"])
+            num_train_eval += num_valid
+            if num_valid > 0 and num_valid < g.number_of_nodes(ntype):
+                balance_arr += num_valid * balance_tag
+                balance_tag += 1
+            logging.debug("Balance %d validation nodes on node %s.", num_valid, ntype)
+        if "test_mask" in node_data[ntype] and num_train_eval < g.number_of_nodes(ntype):
+            num_test = th.sum(node_data[ntype]["test_mask"])
+            num_train_eval += num_test
+            if num_test > 0 and num_test < g.number_of_nodes(ntype):
```
Comment (classicsong) [inline python/graphstorm/gconstruct/utils.py:964]: Shouldn't it be: 
```suggestion
            if num_test > 0 and num_train_eval < g.number_of_nodes(ntype):
```

**Review by classicsong — COMMENTED** [review]
Can you add some unit tests to cover the corner cases?

File: python/graphstorm/gconstruct/utils.py (hunk 1 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/utils.py:964]: this is not a typo. i have test of the number of train/eval nodes in the end.

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 2]
```diff
@@ -1140,6 +1135,67 @@ def test_partition_graph(num_parts):
...
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        partition_graph(g, node_data, edge_data, 'test', num_parts, tmpdirname,
+                        part_method="metis", save_mapping=True)
+        for i in range(num_parts):
+            part_dir = os.path.join(tmpdirname, "part" + str(i))
+            node_data1 = dgl.data.utils.load_tensors(os.path.join(part_dir, 'node_feat.dgl'))
+            assert th.sum(node_data1['node1/train_mask']) > 0
+            assert th.sum(node_data1['node1/test_mask']) == 0
+
+    # case 4: subset are training nodes but not test nodes.
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1169]:
```suggestion
    # case 4: subset are test nodes but not training nodes.
```

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 3]
```diff
@@ -1140,6 +1135,67 @@ def test_partition_graph(num_parts):
...
+
+    # case 5: all nodes are training nodes and test nodes.
+    train_mask1 = np.ones((num_nodes['node1'],), dtype=np.int8)
+    test_mask1 = np.ones((num_nodes['node1'],), dtype=np.int8)
+    node_data['node1']['train_mask'] = train_mask1
+    node_data['node1']['test_mask'] = test_mask1
+    # verify that the graph partitioning can balance the training set.
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        try:
+            partition_graph(g, node_data, edge_data, 'test', num_parts, tmpdirname,
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1193]: Are we checking the case `num_train_eval > g.number_of_nodes(ntype)`?

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 4]
```diff
@@ -1140,6 +1135,67 @@ def test_partition_graph(num_parts):
...
+            node_data1 = dgl.data.utils.load_tensors(os.path.join(part_dir, 'node_feat.dgl'))
+            assert th.sum(node_data1['node1/train_mask']) == 0
+            assert th.sum(node_data1['node1/test_mask']) > 0
+
+    # case 5: all nodes are training nodes and test nodes.
+    train_mask1 = np.ones((num_nodes['node1'],), dtype=np.int8)
+    test_mask1 = np.ones((num_nodes['node1'],), dtype=np.int8)
+    node_data['node1']['train_mask'] = train_mask1
+    node_data['node1']['test_mask'] = test_mask1
+    # verify that the graph partitioning can balance the training set.
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1190]:
```suggestion
    # verify that the graph partitioning will fail.
```

File: tests/unit-tests/gconstruct/test_construct_graph.py [hunk 5]
```diff
@@ -1140,6 +1135,67 @@ def test_partition_graph(num_parts):
...
+    with tempfile.TemporaryDirectory() as tmpdirname:
+        partition_graph(g, node_data, edge_data, 'test', num_parts, tmpdirname,
+                        part_method="metis", save_mapping=True)
+        for i in range(num_parts):
+            part_dir = os.path.join(tmpdirname, "part" + str(i))
+            node_data1 = dgl.data.utils.load_tensors(os.path.join(part_dir, 'node_feat.dgl'))
+            assert th.sum(node_data1['node1/train_mask']) == 0
+            assert th.sum(node_data1['node1/test_mask']) > 0
+
```
Comment (classicsong) [inline tests/unit-tests/gconstruct/test_construct_graph.py:1184]: There is another case: nodes are split into train/valid/test.

File: tests/unit-tests/gconstruct/test_construct_graph.py (hunk 3 above)
Comment (zheng-da) [reply tests/unit-tests/gconstruct/test_construct_graph.py:1210]: yes

**Review by classicsong — APPROVED** [review]
